// @intelligo.ai/bonfire-store/dist/esm2015/lib/bonfire-store.model.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bonfire-store.model.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { WithObservable } from '@intelligo.ai/bonfire';
import { InjectionToken } from '@angular/core';
import { map, tap } from 'rxjs/operators';
// tslint:disable-next-line: no-namespace
export var BonfireStore;
(function (BonfireStore) {
    /**
     * @record
     * @template T
     */
    function Config() { }
    BonfireStore.Config = Config;
    BonfireStore.BONFIRE_CONFIG = new InjectionToken('BONFIRE_CONFIG');
    /**
     * @template T
     */
    class Store {
        /**
         * @param {?} initialValue
         * @param {?=} name
         */
        constructor(initialValue, name) {
            this.name = name;
            if (initialValue) {
                this.store = initialValue;
            }
        }
        /**
         * @param {?=} key
         * @param {?=} callback
         * @return {?}
         */
        get$(key, callback) {
            /** @type {?} */
            const mapCallback = (/**
             * @param {?} store
             * @return {?}
             */
            store => key ? store[(/** @type {?} */ (key))] : store);
            return this.store$
                .asObservable()
                .pipe(map(mapCallback), tap(callback || ((/**
             * @param {?} value
             * @return {?}
             */
            (value) => { }))));
        }
        /**
         * @template J
         * @param {?} key
         * @param {?} value
         * @return {?}
         */
        set$(key, value) {
            this.store[key] = (/** @type {?} */ (value));
        }
        /**
         * @param {?} key
         * @return {?}
         */
        _get(key) {
            return this.store[key];
        }
    }
    tslib_1.__decorate([
        WithObservable(),
        tslib_1.__metadata("design:type", Object)
    ], Store.prototype, "store", void 0);
    BonfireStore.Store = Store;
    if (false) {
        /** @type {?} */
        Store.prototype.store;
        /**
         * @type {?}
         * @private
         */
        Store.prototype.store$;
        /** @type {?} */
        Store.prototype.name;
    }
    /**
     * @record
     * @template J
     */
    function SubStores() { }
    BonfireStore.SubStores = SubStores;
})(BonfireStore || (BonfireStore = {}));
