// @intelligo.ai/bonfire-store/dist/esm5/lib/bonfire-store.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/bonfire-store.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { BonfireStoreComponent } from './bonfire-store.component';
import { BonfireStore } from './bonfire-store.model';
// @dynamic
var BonfireStoreModule = /** @class */ (function () {
    function BonfireStoreModule() {
    }
    /**
     * @param {?=} useValue
     * @return {?}
     */
    BonfireStoreModule.forRoot = /**
     * @param {?=} useValue
     * @return {?}
     */
    function (useValue) {
        if (useValue === void 0) { useValue = {}; }
        return {
            ngModule: BonfireStoreModule,
            providers: [
                { provide: BonfireStore.BONFIRE_CONFIG, useValue: useValue }
            ],
        };
    };
    BonfireStoreModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [BonfireStoreComponent],
                    imports: [],
                    exports: [BonfireStoreComponent]
                },] }
    ];
    return BonfireStoreModule;
}());
export { BonfireStoreModule };
